### PR TITLE
Screen Space Ambient Occlusion Implementation

### PR DESCRIPTION
**Pull request by NEDJIMAbelgacem**
## Description
This PR implement screen space ambient occlusion algorithm into the 3D view frame graph.
Without SSAO:
without-ssao
With SSAO:
ambient-occlusion

The parameter for the SSAO are modifiable under the advanced settings of the 3D view (but hold decent default values for now):

Screenshot from 2022-06-30 13-07-58

### CODE REVIEW

**Comment by benoitdm-oslandia** [comment]
Hi @NEDJIMAbelgacem !

Nice feature! 

Can you elaborate in bit more in the commit message and add some comments to have a better understanding of this feature algorithm (how it works, etc.)?

Regards.

**Comment by nyalldawson** [comment]
Nice! 😁

Have you got any screenshots of how this looks with vector features?

**Comment by wonder-sk** [comment]
@NEDJIMAbelgacem in addition to the bits we have discussed (blurry look of output, unwanted silhouettes, not enough depth perception compared to the results from CloudCompare), I have come across few more things:

- when used in my other testing project (DEM terrain + vector buildings), there seem to be no ambient occlusion applied - the whole scene just gets darker:
  - without SSAO:
    image
  - with SSAO:
    image

- it would be better to use just "ambient occlusion" everywhere in the GUI - the fact it is screen space ambient occlusion is a technical detail and making the text unnecessarily too long

- it would be good to document in the PR description what are likely sensible ranges of the parameters, and possibly when would one want to adjust them

- in the configuration GUI, the whole SSAO group box has some extra margins on the left and right compared to other group boxes

**Comment by NEDJIMAbelgacem** [comment]
> Nice! grin
> 
> Have you got any screenshots of how this looks with vector features?

Unfortunately with vector features the effect is not great.
Here is what it looks like in my testing:
Screenshot from 2022-06-30 13-20-02

File: src/ui/3d/ambientocclusionsettingswidget.ui [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+     <layout class="QGridLayout" name="gridLayout">
+      <item row="2" column="0">
+       <widget class="QLabel" name="label_2">
+        <property name="text">
+         <string>Distance attenuation factor</string>
+        </property>
+       </widget>
+      </item>
+      <item row="3" column="1">
+       <widget class="QDoubleSpinBox" name="mRadiusParameterSpinBox">
```
Comment (uclaros) [inline src/ui/3d/ambientocclusionsettingswidget.ui:46]: A step value of `0.01` would make sense for this.
Also probably a minimum value that is not `0` (`0.001`?)

File: src/ui/3d/ambientocclusionsettingswidget.ui [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+     <property name="title">
+      <string>Ambient Occlusion</string>
+     </property>
+     <property name="checkable">
+      <bool>true</bool>
+     </property>
+     <property name="checked">
+      <bool>false</bool>
+     </property>
+     <layout class="QGridLayout" name="gridLayout">
```
Comment (uclaros) [inline src/ui/3d/ambientocclusionsettingswidget.ui:37]: Tab order within this layout is messed, items should probably be ordered by their row

File: src/3d/qgsshadowrenderingframegraph.h [hunk 3]
```diff
@@ -220,6 +311,13 @@ class QgsShadowRenderingFrameGraph : public Qt3DCore::QEntity
     float mShadowBias = 0.00001f;
     int mShadowMapResolution = 2048;
 
+    // Ambient occlusion related settings
+    bool mAmbientOcclusionEnabled = false;
+    bool mAmbientOcclusionBlurEnabled = true;
+    float mAmbientOcclusionShadingFactor = 300.0f;
```
Comment (uclaros) [inline src/3d/qgsshadowrenderingframegraph.h:317]: This is not initialized to the default value of `50`, does it matter?

File: src/3d/qgsambientocclusionblurentity.cpp [hunk 4]
```diff
@@ -0,0 +1,129 @@
+/***************************************************************************
+  qgsambientocclusionblurentity.cpp
```
Comment (uclaros) [inline src/3d/qgsambientocclusionblurentity.cpp:2]: This class looks a lot like `qgsambientocclusionrenderentity.cpp` . Would it make sense to have the common bits on a parent class?

**Review by uclaros — COMMENTED** [review]
Nice effect! Great work @NEDJIMAbelgacem !

**Comment by nyalldawson** [comment]
> Unfortunately with vector features the effect is not great.

Can you clarify why this is? Everything I've read about regarding AO indicates that it should improve the quality of rendering of pure vector features too. Is it because the implementation is targeted for point clouds specifically?

(If so, I'd say that that's backwards, and that we should prioritise the vector results of AO vs point cloud)

File: src/ui/3d/ambientocclusionsettingswidget.ui [hunk 5]
```diff
@@ -35,58 +35,64 @@
...
+         <string>Radius parameter</string>
         </property>
        </widget>
       </item>
-      <item row="0" column="0" colspan="2">
-       <widget class="QCheckBox" name="mAmbientOcclusionBlurCheckbox">
-        <property name="text">
-         <string>Enable Ambient Occlusion blurring</string>
+      <item row="3" column="1">
+       <widget class="QgsDoubleSpinBox" name="mRadiusParameterSpinBox">
```
Comment (uclaros) [inline src/ui/3d/ambientocclusionsettingswidget.ui:84]: Need to also set the clearValue

**Comment by wonder-sk** [comment]
@nyalldawson 

> Can you clarify why this is? Everything I've read about regarding AO indicates that it should improve the quality of rendering of pure vector features too. Is it because the implementation is targeted for point clouds specifically?

Let me explain - the SSAO effect should work with any kind of 3D data we display on the scene - independent from whether it is point cloud, vector, raster or mesh. For example, here's a test scene with raster-based terrain and vector data, and it's easy to see the visual improvement due to the ambient occlusion:

| SSAO disabled        | SSAO enabled |
| ------------- | ------------- |
| 3d-basic-ssao-off      | 3d-basic-ssao-on |


If SSAO applied to a particular scene does not give visual improvements, there are likely other factors to blame - for example the camera's near and far plane may be too far away from each other, lowering the depth buffer resolution and making the effect less prononouced. Or maybe the SSAO parameters need a bit of tweaking.

Few more implementation notes:
- it is worth reading this article to get a basic understanding of how the effect works: https://learnopengl.com/Advanced-Lighting/SSAO
- in QGIS implementation, for each pixel we sample neighbors from full unit sphere to estimate how is the pixel occluded (https://learnopengl.com/img/advanced-lighting/ssao_crysis_circle.png)
- the link above goes on to discuss a slightly better approach using only hemispheres for estimation, and using normal vectors for each pixel (https://learnopengl.com/img/advanced-lighting/ssao_hemisphere.png). We don't do that in QGIS for two reasons:
   1. we do not keep information about normals for each pixel - this would require update of all existing materials (shaders) to get that information (that would mean also forking materials from Qt3D extras(!))
   2. even if we had information about normals for each pixel, for point clouds we still wouldn't have normals, so we would need to fall back to the method we're using now with sampling from the full sphere.
- in QGIS, we have one post-processing pass (taking depth buffer and color buffer) that estimates ambient occlusion, and then optionally bilateral filtering (blurring) - if enabled - to smooth out artifacts
- QGIS implementation is based on CloudCompare's implementation, with few differences:
  - CloudCompare also uses a texture with random vectors for each pixel to lower the amount of artifacts, blurring is always enabled
  - our creation of samples for the unit sphere is slightly different (CC uses Sobol sequence for more even distribution of samples)
  - CC's implementation only works with orthographic projection, ours only with perspective projection

For those interested, here are different outputs one can get from CC, using different methods. You may need to click on the pictures to see the subtle differences. These options are not user facing, I had to alter CC code+recompile to try them. QGIS should be giving similar results to the noise texture off (with or without blurring)
 - noise texture off, blurring off
    ssao-noise-0-blur-0-gray
 - noise texture on, blurring off
    ssao-noise-1-blur-0-gray
  - noise texture off, blurring on
    ssao-noise-0-blur-1-gray
  - noise texture on, blurring on (this is what CC does normally)
    ssao-noise-1-blur-1-gray

I believe that the results from CC will have slightly different shading also because CC uses orthographic projection, while in QGIS 3D we primarily use perspective projection (support for orthographic projection in QGIS is currently half-broken and hard to maintain).

For future improvements in the QGIS implementation it may be useful to play with:
- addition of the noise texture like in CC (using a random unit vector for each pixel, for fewer artifacts)
- using different method of generating unit sphere samples (may have very little impact though)
- look into how QGIS 3D adjusts near and far planes, to better use range of the depth buffer, alternatively look into if we can make the SSAO effect less dependent on the configuration of camera's near and far planes

**Comment by nyalldawson** [comment]
@wonder-sk 

I'm aware how SSAO should affect vector features -- my reply was in response to @NEDJIMAbelgacem 's comment that the results with vector features are "not great". To me that's indicative of a bug in the implementation, because the results of rendering vector features with SSAO vs without should be a big improvement (and not a degradation of the render)

 Looking at your reference images, there's parts of the scene where I'd expect AO to be occurring but it's not:

image

Specifically there's sharp inner corners where no AO seems to be calculated (there's similar missing bits in @NEDJIMAbelgacem 's image). It's odd because the same situation which is occurring when a polygon intersects the terrain surface at a sharp angle IS resulting in AO...

**Comment by DelazJ** [comment]
> Can you elaborate in bit more in the commit message and add some comments to have a better understanding of this feature algorithm (how it works, etc.)?

@NEDJIMAbelgacem Thanks for having addressed this request... on a developer/expert point of view.
Can you (or someone else) update the **first** message of the PR with an end-user oriented text, please? A text that can easily be reused to explain the feature in the user manual as well as for the changelog. Thanks.

**Comment by nyalldawson** [comment]
@wonder-sk 

> In general, are you happy if this would get merged with the current visual results (pending review), or is this a blocker for you?

I've just given this branch a test run. While it works nicely for point cloud layers, I don't see **any** difference for other features types except that the whole scene gets darker.

Here's an example

Peek 2022-07-15 10-20

**Comment by wonder-sk** [comment]
@nyalldawson I had the same issue initially with that dataset - it is caused by the fact that some chunks of the buildings layer get infinite(?) z range when there's no data in the chunk, which causes very bad near+far plane setup:

image

While this is a separate issue to be fixed, a quick workaround is to change the "zoom levels count" property from 3 to 1, to get just a single tile with correct z range, and the near/far plane get set to reasonable values and the effect starts to work:

image

**Comment by nyalldawson** [comment]
@wonder-sk 

> While this is a separate issue to be fixed, a quick workaround is to change the "zoom levels count" property from 3 to 1, to get just a single tile with correct z range, and the near/far plane get set to reasonable values and the effect starts to work:

Ok, I'll try that, thanks!

>  it is caused by the fact that some chunks of the buildings layer get infinite(?) z range when there's no data in the chunk, which causes very bad near+far plane setup:

So I guess in the situation that a vector layer chunk is empty we should completely remove that node, shouldn't we? I can't see anything in the api which currently lets us set a node as empty and to never attempt to load it again. Am I missing something?

**Comment by wonder-sk** [comment]
> So I guess in the situation that a vector layer chunk is empty we should completely remove that node, shouldn't we? I can't see anything in the api which currently lets us set a node as empty and to never attempt to load it again. Am I missing something?

Yeah that seems like the right thing to do. Actually there is such a mechanism already, working currently for DEM terrain (when there are no data values) - in QgsChunkedEntity::onActiveJobFinished() if createEntity() returns null pointer (because no data were found), the node will be marked with `node->setHasData( false );` and won't be loaded again. We just need to handle that in QgsVectorLayerChunkLoader::createEntity() and QgsRuleBasedChunkLoader::createEntity() as well.

**Comment by nyalldawson** [comment]
@wonder-sk 

How's https://github.com/qgis/QGIS/pull/49375 look to you?

**Comment by nyalldawson** [comment]
@wonder-sk 

I retested with #49375 in place. Unfortunately still no luck (although the results are a bit different).

Here's without SSAO

image

And with SSAO

image

So the whole scene is a bit darker, there's some weird blocky artefacts on the terrain surface, and possibly some inverted(?) brighter highlighting on the edges of polygons. :thinking:

**Comment by nyalldawson** [comment]
A better demo: The artefacts aren't a result of GIF compression -- this is how it actually looks on my screen!
Peek 2022-07-18 08-26

**Comment by wonder-sk** [comment]
@nyalldawson  Try changing the radius to a lower value, from the current default 0.05 to something like 0.01, that will make the shading bleed out much less. The value 0.05 seems a bit too large (if I interpret the parameter correctly, the neighborhood of a pixel where sampling is done would be 5% of image width/height).

The brighter highlight on the edges of polygons is something that can be expected (I think) using this kind of implementation: near the edges, most of the sampled neighbors are in front of the camera, so there's very little to no darkening applied compared to other parts of the scene. Lowering the radius would also make this less pronounced...

**Comment by nyalldawson** [comment]
@wonder-sk 

That helps a bit, but the results still just seem wrong to me:

image

Here I'm **only** seeing the occlusion happening as a kind of "outer glow" around the top edges of the features. That's totally wrong to me... there's nothing there to be occluded! And then in the places where the occlusion *should* be visible, we see no change at all (the parts were the feature meets the terrain, and the interior corner of the feature). So it honestly looks to me like the whole effect is being inverted somehow...

**Comment by nyalldawson** [comment]
That said, the occlusion of the terrain with itself looks fine :confused: 

image

**Comment by wonder-sk** [comment]
Replaced by #49702

**Pull request closed without merging**